### PR TITLE
implemented rank of a matrix via SVD as mentioned in issue #608

### PR DESCRIPTION
**Pull request by pgtgrly**
I have implemented the function to get rank via SVD as svd_rank() in math.rb. I found an abandoned PR [here](https://github.com/SciRuby/nmatrix/pull/461) and updated the code for ShapeError,user-defined tolerance along with the required documentation.

Ref #608.

### CODE REVIEW

**Comment by v0dro** [comment]
Two things:
* Write tests for all features that you add.
* Make the build pass.

**Comment by pgtgrly** [comment]
I will get to it!

**Comment by mohawkjohn** [comment]
It may not be possible for the build to pass all tests. There are some problems with the Travis configuration. If you see any error messages that don't make any sense for this particular problem, you can probably ignore them; if you're unsure, please ask. But yes, Sameer is correct that you need to write tests, and those need to pass. :)

**Comment by pgtgrly** [comment]
@v0dro @mohawkjohn 
Greetings,
I checked the log for Travis build files and the errors are not related to the code that I wrote. In [one case](https://travis-ci.org/SciRuby/nmatrix/jobs/360010680) the homebrew:science tap is deprecated, in [another](https://travis-ci.org/SciRuby/nmatrix/jobs/360010681) a log file is missing and rest require a Y/N prompt. 

I have added certain tests to math_spec.rb for svd_rank() based on the examples found [here](https://www.cliffsnotes.com/study-guides/algebra/linear-algebra/real-euclidean-vector-spaces/the-rank-of-a-matrix). Please let me know if further tests are needed. I have also included nmatrix/lapacke in spec_helper.rb as svd_rank requires nmatrix-lapacke. 

On a side note, I think it should be explicitly stated in the Developer guide that rspec v 2.99 is used as [LongRun is not implemented in v3.x](https://github.com/mdub/rspec-longrun/issues/5)

**Comment by pgtgrly** [comment]
I removed "require nmatrix/lapacke " from spec_helper to complete the builds and put a "rescue NotImplementedError". However, it is leading to a nomethod error in this [build](https://travis-ci.org/SciRuby/nmatrix/jobs/360972275) (using jruby).

File: spec/math_spec.rb [hunk 1]
```diff
@@ -1287,43 +1287,58 @@
...
           begin
             rank = a.svd_rank()
-          end
+          
+            rank_true = 3
+            expect(rank).to eq (rank_true)
 
-          rank_true = 3
-          expect(rank).to eq (rank_true)
+          rescue NotImplementedError
```
Comment (mohawkjohn) [inline spec/math_spec.rb]: Okay, interesting approach here. Generally if a method isn't supported under a user's specific configuration, we don't use rescue in specs. We just check with an `if` statement and use `pending` as you've done; we just do it without the `rescue`. Can you look around a bit and make it consistent with how we've done it elsewhere?

File: spec/spec_helper.rb [hunk 2]
```diff
@@ -28,7 +28,6 @@
 #require "narray/narray"
 
 require "./lib/nmatrix"
-require "./lib/nmatrix/lapacke"
```
Comment (mohawkjohn) [inline spec/spec_helper.rb]: Why did you remove this specific line?

File: lib/nmatrix/math.rb [hunk 3]
```diff
@@ -725,7 +725,7 @@ def svd_rank(tolerence="default")
       when "default"
         tolerence = self.shape.max * sigmas.max * eps # tolerence of a Matrix A is max(size(A))*eps(norm(A)). norm(A) is nearly equal to max(sigma of A)
     end
-    sigmas.map { |x| x > tolerence ? 1 : 0 }.reduce(:+)
+    return sigmas.map { |x| x > tolerence ? 1 : 0 }.reduce(:+)
```
Comment (mohawkjohn) [inline lib/nmatrix/math.rb]: Explicit returns are not needed in Ruby.

File: spec/spec_helper.rb (hunk 2 above)
Comment (mohawkjohn) [reply spec/spec_helper.rb]: Ah! I see. You added it in an earlier commit.

**Comment by mohawkjohn** [comment]
> On a side note, I think it should be explicitly stated in the Developer guide that rspec v 2.99 is used as LongRun is not implemented in v3.x.

Where did this come up for you? In the gemspec, NMatrix requires rspec 2.14.

**Comment by mohawkjohn** [comment]
Can you please do a squash of all of these commits when you get a chance? Makes them much easier to review.

**Comment by pgtgrly** [comment]
Greetings John,
I have squashed my commits together and I see that it is much cleaner and easier to read.  I will make sure to do this in future situations. I tried to run rspec outside without rake and that gave me the error mentioned [here](https://github.com/mdub/rspec-longrun/issues/5) would have been helpful if it was mentioned in the developer's guide because of the Test Driven Development of the project.

Like before, I checked the log for Travis build files and the errors are not related to the code that I wrote. In one case the homebrew:science tap is deprecated, in another a log file is missing and rest require a Y/N prompt.

File: ext/nmatrix/nmatrix.cpp [hunk 4]
```diff
@@ -32,6 +32,7 @@
  */
 
 #include <ruby.h>
+#include <float.h>
```
Comment (mohawkjohn) [inline ext/nmatrix/nmatrix.cpp]: In C++ code, please don't do `<float.h>`, do `<cfloat>`.

**Comment by pgtgrly** [comment]
Greetings @mohawkjohn, I have made the required changes. I request you to please review my PR and let me know if something needs to be changed :)
